// @clevercloud/client/esm/api/v4/materia.js
import { pickNonNull } from '../../pick-non-null.js';

/**
 * POST /iam/add-allowed-owner
 * @param {Object} params
 * @param {String} params.undefined
 */
export function addAllowedOwner(params) {
  // no multipath for /self or /organisations/{id}
  return Promise.resolve({
    method: 'post',
    url: `/v4/iam/add-allowed-owner`,
    headers: { Accept: 'application/json' },
    // no query params
    // no body
  });
}

/**
 * GET /iam/organisations/{ownerId}/materia-kv/{materiaId}/api-keys
 * @param {Object} params
 * @param {String} params.undefined
 * @param {String} params.undefined
 * @param {String} params.from
 * @param {String} params.size
 */
export function listMateriaKvApiKeys(params) {
  // no multipath for /self or /organisations/{id}
  return Promise.resolve({
    method: 'get',
    url: `/v4/iam/organisations/${params.ownerId}/materia-kv/${params.materiaId}/api-keys`,
    headers: { Accept: 'application/json' },
    queryParams: pickNonNull(params, ['from', 'size']),
    // no body
  });
}

/**
 * POST /iam/organisations/{ownerId}/materia-kv/{materiaId}/api-keys
 * @param {Object} params
 * @param {String} params.undefined
 * @param {String} params.resource
 * @param {String} params.expirationdate
 */
export function createMateriaKvApiKey(params) {
  // no multipath for /self or /organisations/{id}
  return Promise.resolve({
    method: 'post',
    url: `/v4/iam/organisations/${params.ownerId}/materia-kv/${params.materiaId}/api-keys`,
    headers: { Accept: 'application/json' },
    queryParams: pickNonNull(params, ['resource', 'expirationdate']),
    // no body
  });
}

/**
 * DELETE /iam/organisations/{ownerId}/materia-kv/{materiaId}/api-keys/{apiKeyId}
 * @param {Object} params
 * @param {String} params.undefined
 * @param {String} params.undefined
 * @param {String} params.undefined
 */
export function revokeMateriaKvApiKey(params) {
  // no multipath for /self or /organisations/{id}
  return Promise.resolve({
    method: 'delete',
    url: `/v4/iam/organisations/${params.ownerId}/materia-kv/${params.materiaId}/api-keys/${params.apiKeyId}`,
    headers: {},
    // no query params
    // no body
  });
}

/**
 * GET /iam/organisations/{ownerId}/materia-kv/{materiaId}/api-keys/{apiKeyId}
 * @param {Object} params
 * @param {String} params.undefined
 * @param {String} params.undefined
 * @param {String} params.undefined
 */
export function getMateriaKvApiKey(params) {
  // no multipath for /self or /organisations/{id}
  return Promise.resolve({
    method: 'get',
    url: `/v4/iam/organisations/${params.ownerId}/materia-kv/${params.materiaId}/api-keys/${params.apiKeyId}`,
    headers: { Accept: 'application/json' },
    // no query params
    // no body
  });
}
